compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,409 @@
+    if !t.is_empty() {
+        s.push_str(&t);
+    } else {
+        s.push('v')
+    }
+
+    let tys = fn_sig.inputs();
+    if !tys.is_empty() {
+        // Encode the parameter types
+        for ty in tys {
+            s.push_str(&encode_ty(tcx, *ty, dict, encode_c_types));
+        }
+
+        if fn_sig.c_variadic {
+            s.push('z');
+        }
+    } else {
+        if fn_sig.c_variadic {
+            s.push('z');
+        } else {
+            // Empty parameter lists, whether declared as () or conventionally as (void), are
+            // encoded with a void parameter specifier "v".
+            s.push('v')
+        }
+    }
+
+    // Close the "F..E" pair
+    s.push('E');
+
+    s.to_string()
+}
+
+// Encodes a ty:Ty using the Itanium C++ ABI with vendor extended type qualifiers and types for
+// Rust types that are not used at the FFI boundary.
+fn encode_ty<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    ty: Ty<'tcx>,
+    dict: &mut FxHashMap<Ty<'tcx>, usize>,
+    encode_c_types: bool,
+) -> String {
+    let mut typeid = String::new();
+
+    match ty.kind() {
+        // Primitive types
+        ty::Bool | ty::Int(_) | ty::Uint(_) | ty::Float(_) => {
+            let mut s = String::from(match ty.kind() {
+                // u<type-name> as vendor extended type
+                ty::Bool => "b",
+                ty::Int(IntTy::I8) => "u2i8",
+                ty::Int(IntTy::I16) => "u3i16",
+                ty::Int(IntTy::I32) => "u3i32",
+                ty::Int(IntTy::I64) => "u3i64",
+                ty::Int(IntTy::I128) => "u4i128",
+                ty::Int(IntTy::Isize) => "u5isize",
+                ty::Uint(UintTy::U8) => "u2u8",
+                ty::Uint(UintTy::U16) => "u3u16",
+                ty::Uint(UintTy::U32) => "u3u32",
+                ty::Uint(UintTy::U64) => "u3u64",
+                ty::Uint(UintTy::U128) => "u4u128",
+                ty::Uint(UintTy::Usize) => "u5usize",
+                ty::Float(FloatTy::F32) => "u3f32",
+                ty::Float(FloatTy::F64) => "u3f64",
+                _ => "",
+            });
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Char => {
+            // u<type-name> as vendor extended type
+            let mut s = String::from("u4char");
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Str => {
+            // u<type-name> as vendor extended type
+            let mut s = String::from("u3str");
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::Never => {
+            // u<type-name> as vendor extended type
+            let mut s = String::from("u5never");
+            compress(dict, ty, &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Compound types
+        // () in Rust is equivalent to void return type in C
+        ty::Tuple(_) if ty.is_unit() => {}
```

Comment:
It seems surprising to use empty mangling for `()`, that wouldn't be valid according to the grammar in general.

For example, if it is necessary to skip `()` return type in a function signature, this special case probably belongs somewhere else.

Reply:
It's actually encoded as 'v' if encode_ty returns an empty string. This code changed a lot throughout multiple iterations, and I should have moved it here. Done.